### ISSUE TITLE
sql: ROWS FROM functions not finding columns that should be in scope

### ISSUE THREAD

**Issue by andrioni** [MaterializeInc/materialize#9653]
### What version of Materialize are you using?

From the shell:

```
 v0.12.1-dev (42ad74326)
```

### How did you install Materialize?

* [x] Built from source

### What was the issue?
The following query works in Postgres, but fails in Materialize:
```sql
-- Postgres
alessandro=# select indx, val
from foo,
     rows from(generate_series(1, jsonb_array_length(foo.f)), 
               jsonb_array_elements(foo.f)) r(indx, val);

 indx |                 val
------+--------------------------------------
    1 | {"type": "standard", "title": "bar"}
    2 | {"type": "standard", "title": "baz"}
(2 rows)
```

```sql
-- Materialize
select indx, val
from foo,
     rows from(generate_series(1, jsonb_array_length(foo.f)), 
               jsonb_array_elements(foo.f)) r(indx, val);
ERROR:  column "foo.f" does not exist
```

If I remove the `ROWS FROM` clause, then it correctly finds `foo.f`, but we lose the effects of `ROWS FROM`:
```sql
materialize=> select indx, val
from foo,
     generate_series(1, jsonb_array_length(foo.f)) indx,
     jsonb_array_elements(foo.f) val;
 indx |                val
------+-----------------------------------
    1 | {"title":"bar","type":"standard"}
    1 | {"title":"baz","type":"standard"}
    2 | {"title":"bar","type":"standard"}
    2 | {"title":"baz","type":"standard"}
(4 rows)
```

This issue was motivated by [this conversation](https://materializecommunity.slack.com/archives/C015KDVS7EV/p1639703972312000) on Community Slack.

### Is the issue reproducible? If so, please provide reproduction instructions.
```sql
create table foo (f jsonb);
insert into foo values ('[{"title":"bar","type":"standard"}, {"title":"baz","type":"standard"}]');
```

**Comment by asenac** [comment]
It panics in latest main `(materialized v0.12.1-dev (520313a87))`: 
```
Thread 25 "coordinator" hit Breakpoint 1, materialized::handle_panic (panic_info=0x7fffebe1dad8) at src/materialized/src/bin/materialized/main.rs:789
789         let _guard = PANIC_MUTEX.lock();
(gdb) bt
#0  materialized::handle_panic (panic_info=0x7fffebe1dad8) at src/materialized/src/bin/materialized/main.rs:789
#1  0x0000555557c74978 in core::ops::function::Fn::call () at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/core/src/ops/function.rs:70
#2  0x000055555df07de9 in std::panicking::rust_panic_with_hook () at library/std/src/panicking.rs:628
#3  0x000055555df07842 in std::panicking::begin_panic_handler::{{closure}} () at library/std/src/panicking.rs:519
#4  0x000055555df05a34 in std::sys_common::backtrace::__rust_end_short_backtrace () at library/std/src/sys_common/backtrace.rs:139
#5  0x000055555df077d9 in rust_begin_unwind () at library/std/src/panicking.rs:517
#6  0x000055555df2c591 in core::panicking::panic_fmt () at library/core/src/panicking.rs:100
#7  0x000055555df2c47d in core::panicking::panic () at library/core/src/panicking.rs:50
#8  0x000055555b5feb98 in sql::plan::lowering::<impl sql::plan::expr::HirRelationExpr>::applied_to::{{closure}} () at src/sql/src/plan/lowering.rs:309
#9  0x000055555b8917ea in stacker::maybe_grow (red_zone=262144, stack_size=16777216, callback=...) at /home/asenac/.cargo/registry/src/github.com-1ecc6299db9ec823/stacker-0.1.14/src/lib.rs:55
#10 ore::stack::maybe_grow (f=...) at /home/asenac/work/asenac-materialize/src/ore/src/stack.rs:100
#11 sql::plan::lowering::<impl sql::plan::expr::HirRelationExpr>::applied_to (self=..., id_gen=0x7fffebe33b90, get_outer=..., col_map=0x7fffebe28c00) at src/sql/src/plan/lowering.rs:162
#12 0x000055555b600a76 in sql::plan::lowering::<impl sql::plan::expr::HirRelationExpr>::applied_to::{{closure}} () at src/sql/src/plan/lowering.rs:188
#13 0x000055555b8917ea in stacker::maybe_grow (red_zone=262144, stack_size=16777216, callback=...) at /home/asenac/.cargo/registry/src/github.com-1ecc6299db9ec823/stacker-0.1.14/src/lib.rs:55
#14 ore::stack::maybe_grow (f=...) at /home/asenac/work/asenac-materialize/src/ore/src/stack.rs:100
#15 sql::plan::lowering::<impl sql::plan::expr::HirRelationExpr>::applied_to (self=..., id_gen=0x7fffebe33b90, get_outer=..., col_map=0x7fffebe28c00) at src/sql/src/plan/lowering.rs:162
#16 0x000055555b5f9ad2 in sql::plan::lowering::<impl sql::plan::expr::HirRelationExpr>::applied_to::{{closure}}::{{closure}}::{{closure}} (id_gen=0x7fffebe33b90, right=..., get_left=..., col_map=0x7fffebe28c00) at src/sql/src/plan/lowering.rs:321
#17 0x000055555b60ae22 in sql::plan::lowering::branch::{{closure}}::{{closure}} (id_gen=0x7fffebe33b90, get_keyed_outer=...) at src/sql/src/plan/lowering.rs:1176
#18 0x000055555b9c9806 in expr::relation::MirRelationExpr::let_in (self=..., id_gen=0x7fffebe33b90, body=...) at /home/asenac/work/asenac-materialize/src/expr/src/relation/mod.rs:1130
#19 0x000055555b60b8c9 in sql::plan::lowering::branch::{{closure}} (id_gen=0x7fffebe33b90, get_outer=...) at src/sql/src/plan/lowering.rs:1174
#20 0x000055555b9ceb7b in expr::relation::MirRelationExpr::let_in (self=..., id_gen=0x7fffebe33b90, body=...) at /home/asenac/work/asenac-materialize/src/expr/src/relation/mod.rs:1123
#21 0x000055555b608fa5 in sql::plan::lowering::branch (id_gen=0x7fffebe33b90, outer=..., col_map=0x7fffebe331a0, inner=..., apply_requires_distinct_outer=false, apply=...) at src/sql/src/plan/lowering.rs:1162
#22 0x000055555b5f9c7e in sql::plan::lowering::<impl sql::plan::expr::HirRelationExpr>::applied_to::{{closure}}::{{closure}} (id_gen=0x7fffebe33b90, get_left=...) at src/sql/src/plan/lowering.rs:314
#23 0x000055555b9cdce6 in expr::relation::MirRelationExpr::let_in (self=..., id_gen=0x7fffebe33b90, body=...) at /home/asenac/work/asenac-materialize/src/expr/src/relation/mod.rs:1130
#24 0x000055555b5feca2 in sql::plan::lowering::<impl sql::plan::expr::HirRelationExpr>::applied_to::{{closure}} () at src/sql/src/plan/lowering.rs:312
#25 0x000055555b8917ea in stacker::maybe_grow (red_zone=262144, stack_size=16777216, callback=...) at /home/asenac/.cargo/registry/src/github.com-1ecc6299db9ec823/stacker-0.1.14/src/lib.rs:55
#26 ore::stack::maybe_grow (f=...) at /home/asenac/work/asenac-materialize/src/ore/src/stack.rs:100
#27 sql::plan::lowering::<impl sql::plan::expr::HirRelationExpr>::applied_to (self=..., id_gen=0x7fffebe33b90, get_outer=..., col_map=0x7fffebe331a0) at src/sql/src/plan/lowering.rs:162
#28 0x000055555b600a76 in sql::plan::lowering::<impl sql::plan::expr::HirRelationExpr>::applied_to::{{closure}} () at src/sql/src/plan/lowering.rs:188
#29 0x000055555b8917ea in stacker::maybe_grow (red_zone=262144, stack_size=16777216, callback=...) at /home/asenac/.cargo/registry/src/github.com-1ecc6299db9ec823/stacker-0.1.14/src/lib.rs:55
#30 ore::stack::maybe_grow (f=...) at /home/asenac/work/asenac-materialize/src/ore/src/stack.rs:100
#31 sql::plan::lowering::<impl sql::plan::expr::HirRelationExpr>::applied_to (self=..., id_gen=0x7fffebe33b90, get_outer=..., col_map=0x7fffebe331a0) at src/sql/src/plan/lowering.rs:162
#32 0x000055555b5f97b7 in sql::plan::lowering::<impl sql::plan::expr::HirRelationExpr>::lower::{{closure}} (id_gen=0x7fffebe33b90, get_outer=...) at src/sql/src/plan/lowering.rs:138
#33 0x000055555b9ca7f6 in expr::relation::MirRelationExpr::let_in (self=..., id_gen=0x7fffebe33b90, body=...) at /home/asenac/work/asenac-materialize/src/expr/src/relation/mod.rs:1130
#34 0x000055555b891428 in sql::plan::lowering::<impl sql::plan::expr::HirRelationExpr>::lower (self=...) at src/sql/src/plan/lowering.rs:136
#35 0x000055555b89377b in sql::plan::optimize::<impl sql::plan::expr::HirRelationExpr>::optimize_and_lower (self=..., config=0x7fffebe34ac7) at src/sql/src/plan/optimize.rs:55
#36 0x000055555ba2b2db in sql::plan::statement::dml::plan_query (scx=0x7fffebe364d8, query=..., params=0x7fffebe5e688, lifetime=...) at src/sql/src/plan/statement/dml.rs:270
#37 0x000055555ba292f4 in sql::plan::statement::dml::plan_select (scx=0x7fffebe364d8, params=0x7fffebe5e688, copy_to=...) at src/sql/src/plan/statement/dml.rs:152
#38 0x000055555ba1e858 in sql::plan::statement::plan (pcx=..., catalog=..., stmt=..., params=0x7fffebe5e688) at src/sql/src/plan/statement.rs:215
#39 0x00005555580688fe in coord::coord::Coordinator<C>::handle_statement::{{closure}} () at /home/asenac/work/asenac-materialize/src/coord/src/coord.rs:1439
#40 0x0000555557d7078d in <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll (self=..., cx=0x7fffebe5ca08) at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/core/src/future/mod.rs:80
#41 0x0000555557d68dfa in <F as futures_core::future::TryFuture>::try_poll (self=..., cx=0x7fffebe5ca08) at /home/asenac/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-core-0.3.18/src/future.rs:82
#42 0x0000555557f44869 in <futures_util::future::try_future::try_flatten::TryFlatten<Fut,<Fut as futures_core::future::TryFuture>::Ok> as core::future::future::Future>::poll (self=..., cx=0x7fffebe5ca08) at /home/asenac/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-util-0.3.18/src/future/try_future/try_flatten.rs:57
#43 0x0000555557e72486 in <futures_util::future::try_future::TryFlatten<Fut1,Fut2> as core::future::future::Future>::poll (self=..., cx=0x7fffebe5ca08) at /home/asenac/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-util-0.3.18/src/lib.rs:95
#44 0x0000555557e724d6 in <futures_util::future::try_future::AndThen<Fut1,Fut2,F> as core::future::future::Future>::poll (self=..., cx=0x7fffebe5ca08) at /home/asenac/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-util-0.3.18/src/lib.rs:95
#45 0x0000555558082e7a in coord::coord::Coordinator<C>::message_statement_ready::{{closure}} () at /home/asenac/work/asenac-materialize/src/coord/src/coord.rs:833
#46 0x0000555557d6e10b in <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll (self=..., cx=0x7fffebe5ca08) at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/core/src/future/mod.rs:80
#47 0x0000555558094d58 in coord::coord::Coordinator<C>::serve::{{closure}} () at /home/asenac/work/asenac-materialize/src/coord/src/coord.rs:649
#48 0x0000555557d7259b in <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll (self=..., cx=0x7fffebe5ca08) at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/core/src/future/mod.rs:80
#49 0x00005555582a35b0 in tokio::park::thread::CachedParkThread::block_on::{{closure}} () at /home/asenac/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.14.0/src/park/thread.rs:263
#50 0x0000555557d27d80 in tokio::coop::with_budget::{{closure}} (cell=0x7fffebe7affa) at /home/asenac/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.14.0/src/coop.rs:106
#51 0x0000555557d610e8 in std::thread::local::LocalKey<T>::try_with (self=0x55555df47a80, f=...) at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/std/src/thread/local.rs:399
#52 0x0000555557d600dd in std::thread::local::LocalKey<T>::with (self=0x55555df47a80, f=...) at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/std/src/thread/local.rs:375
#53 0x00005555582a2e09 in tokio::coop::with_budget (budget=..., f=...) at /home/asenac/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.14.0/src/coop.rs:99
#54 tokio::coop::budget (f=...) at /home/asenac/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.14.0/src/coop.rs:76
#55 tokio::park::thread::CachedParkThread::block_on (self=0x7fffebe61cd8, f=...) at /home/asenac/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.14.0/src/park/thread.rs:263
#56 0x0000555557c69a07 in tokio::runtime::enter::Enter::block_on (self=0x7fffebe66f48, f=...) at /home/asenac/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.14.0/src/runtime/enter.rs:151
#57 0x0000555557e1a7b2 in tokio::runtime::handle::Handle::block_on (self=0x7fffebe75c10, future=...) at /home/asenac/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.14.0/src/runtime/handle.rs:326
#58 0x000055555809c1ce in coord::coord::serve::{{closure}}::{{closure}} () at /home/asenac/work/asenac-materialize/src/coord/src/coord.rs:4418
#59 0x0000555557d1751f in std::sys_common::backtrace::__rust_begin_short_backtrace (f=...) at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/std/src/sys_common/backtrace.rs:123
#60 0x0000555557d23743 in std::thread::Builder::spawn_unchecked::{{closure}}::{{closure}} () at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/std/src/thread/mod.rs:483
#61 0x0000555557e25b03 in <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once (self=..., _args=()) at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/core/src/panic/unwind_safe.rs:271
#62 0x0000555557f83d60 in std::panicking::try::do_call (data=0x7fffebe76ec8 "\000") at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/std/src/panicking.rs:403
#63 0x0000555557fa5acb in __rust_try ()
#64 0x0000555557f6180d in std::panicking::try (f=...) at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/std/src/panicking.rs:367
#65 0x0000555557f2a423 in std::panic::catch_unwind (f=...) at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/std/src/panic.rs:133
#66 0x0000555557d2190f in std::thread::Builder::spawn_unchecked::{{closure}} () at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/std/src/thread/mod.rs:482
#67 0x0000555557c74ebe in core::ops::function::FnOnce::call_once{{vtable-shim}} () at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/core/src/ops/function.rs:227
#68 0x000055555df0e043 in <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once () at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/alloc/src/boxed.rs:1691
#69 <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once () at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/alloc/src/boxed.rs:1691
#70 std::sys::unix::thread::Thread::new::thread_start () at library/std/src/sys/unix/thread.rs:106
#71 0x00007ffff7f94609 in start_thread (arg=<optimized out>) at pthread_create.c:477
#72 0x00007ffff7d2f293 in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
(gdb) 
```

**Comment by asenac** [comment]
The name resolution issue seems fixed in latest main after #9623. However, the following assert fires during lowering

```
(gdb) frame 8
#8  0x000055555b5feb98 in sql::plan::lowering::<impl sql::plan::expr::HirRelationExpr>::applied_to::{{closure}} () at src/sql/src/plan/lowering.rs:309
309                         assert!(kind.can_be_correlated());
(gdb) print kind
$1 = sql::plan::expr::JoinKind::FullOuter
(gdb) 
```
The SQL plan is:

```
materialize=> explain raw plan for select indx, val
from foo,
     rows from(generate_series(1, jsonb_array_length(foo.f)), 
               jsonb_array_elements(foo.f)) r(indx, val);
                               Raw Plan                               
----------------------------------------------------------------------
 %0 =                                                                +
 | Get materialize.public.foo (u279)                                 +
                                                                     +
 %1 =                                                                +
 | CallTable generate_series(i32toi64(1), jsonb_array_length(#^0), 1)+
 | Map row_number() over ()                                          +
 | Project (#1, #0)                                                  +
                                                                     +
 %2 =                                                                +
 | CallTable jsonb_array_elements(#^0)                               +
 | Map row_number() over ()                                          +
 | Project (#1, #0)                                                  +
                                                                     +
 %3 =                                                                +
 | FullOuterJoin %1 %2 on (#0 = #2)                                  +
 | Project (#1, #3)                                                  +
                                                                     +
 %4 =                                                                +
 | InnerJoin %0 %3 on true                                           +
 | Project (#1, #2)                                                  +
 
(1 row)

materialize=>
```
